File: gommo-backend/src/main/java/br/com/gommo/modules/rh/person/collaborators/admission/mapper/AdmissionProcessMapper.java
package br.com.gommo.modules.rh.person.collaborators.admission.mapper;

import java.util.ArrayList;
import java.util.List;

import org.springframework.stereotype.Component;
import org.springframework.util.CollectionUtils;

import br.com.gommo.core.entity.StatusEnum;
import br.com.gommo.modules.rh.person.collaborators.admission.dto.AdmissionEmergencyContactDto;
import br.com.gommo.modules.rh.person.collaborators.admission.dto.AdmissionProcessRequestDto;
import br.com.gommo.modules.rh.person.collaborators.admission.dto.AdmissionProcessResponseDto;
import br.com.gommo.modules.rh.person.collaborators.admission.dto.AdmissionProgress;
import br.com.gommo.modules.rh.person.collaborators.admission.entity.AdmissionProcess;
import br.com.gommo.modules.rh.person.collaborators.admission.entity.AdmissionStatusEnum;
import br.com.gommo.modules.rh.person.contract.entity.ContractTypeEnum;

@Component
public class AdmissionProcessMapper {

    private static final int REQUIRED_STEP_COUNT = 6;

    public AdmissionProcess toEntity(AdmissionProcessRequestDto dto) {
        return AdmissionProcess.builder()
                .admissionStatus(
                        dto.getAdmissionStatus() != null ? dto.getAdmissionStatus() : AdmissionStatusEnum.IN_PROGRESS)
                .startedAt(dto.getStartedAt())
                .notes(dto.getNotes())
                .fullName(dto.getFullName())
                .socialName(dto.getSocialName())
                .cpf(dto.getCpf())
                .rg(dto.getRg())
                .rgIssuer(dto.getRgIssuer())
                .rgStateCode(dto.getRgStateCode())
                .birthDate(dto.getBirthDate())
                .gender(dto.getGender())
                .maritalStatus(dto.getMaritalStatus())
                .motherName(dto.getMotherName())
                .fatherName(dto.getFatherName())
                .nationality(dto.getNationality() != null ? dto.getNationality() : "Brasileiro")
                .pisPasep(dto.getPisPasep())
                .email(dto.getEmail())
                .phone(dto.getPhone())
                .zipCode(dto.getZipCode())
                .street(dto.getStreet())
                .number(dto.getNumber())
                .complement(dto.getComplement())
                .district(dto.getDistrict())
                .expectedStartDate(dto.getExpectedStartDate())
                .companyId(dto.getCompanyId())
                .departmentId(dto.getDepartmentId())
                .jobPositionId(dto.getJobPositionId())
                .contractType(dto.getContractType() != null ? dto.getContractType() : ContractTypeEnum.CLT)
                .baseSalary(dto.getBaseSalary())
                .workloadSchedule(dto.getWorkloadSchedule())
                .workScheduleId(dto.getWorkScheduleId())
                .emergencyContacts(copyEmergencyContacts(dto.getEmergencyContacts()))
                .contractStartDate(dto.getContractStartDate())
                .contractEndDate(dto.getContractEndDate())
                .providerCnpj(dto.getProviderCnpj())
                .providerLegalName(dto.getProviderLegalName())
                .providerTradeName(dto.getProviderTradeName())
                .recessEnabled(dto.isRecessEnabled())
                .recessTotalDaysPerCycle(dto.getRecessTotalDaysPerCycle())
                .recessCycleMonths(dto.getRecessCycleMonths())
                .recessEligibilityAfterMonths(dto.getRecessEligibilityAfterMonths())
                .recessFinancialMode(dto.getRecessFinancialMode())
                .recessPaidPercentage(dto.getRecessPaidPercentage())
                .recessAllowSplit(dto.isRecessAllowSplit())
                .recessMaxSplitPeriods(dto.getRecessMaxSplitPeriods())
                .recessMinimumSplitDays(dto.getRecessMinimumSplitDays())
                .recessAdvanceNoticeDays(
                        dto.getRecessAdvanceNoticeDays() != null ? dto.getRecessAdvanceNoticeDays() : 0)
                .recessNotes(dto.getRecessNotes())
                .photoObjectId(dto.getPhotoObjectId())
                .build();
    }

    public void updateEntity(AdmissionProcess entity, AdmissionProcessRequestDto dto) {
        entity.setStartedAt(dto.getStartedAt());
        entity.setNotes(dto.getNotes());
        entity.setFullName(dto.getFullName());
        entity.setSocialName(dto.getSocialName());
        entity.setCpf(dto.getCpf());
        entity.setRg(dto.getRg());
        entity.setRgIssuer(dto.getRgIssuer());
        entity.setRgStateCode(dto.getRgStateCode());
        entity.setBirthDate(dto.getBirthDate());
        entity.setGender(dto.getGender());
        entity.setMaritalStatus(dto.getMaritalStatus());
        entity.setMotherName(dto.getMotherName());
        entity.setFatherName(dto.getFatherName());
        if (dto.getNationality() != null) {
            entity.setNationality(dto.getNationality());
        }
        entity.setPisPasep(dto.getPisPasep());
        entity.setEmail(dto.getEmail());
        entity.setPhone(dto.getPhone());
        entity.setZipCode(dto.getZipCode());
        entity.setStreet(dto.getStreet());
        entity.setNumber(dto.getNumber());
        entity.setComplement(dto.getComplement());
        entity.setDistrict(dto.getDistrict());
        entity.setExpectedStartDate(dto.getExpectedStartDate());
        entity.setCompanyId(dto.getCompanyId());
        entity.setDepartmentId(dto.getDepartmentId());
        entity.setJobPositionId(dto.getJobPositionId());
        if (dto.getContractType() != null) {
            entity.setContractType(dto.getContractType());
        }
        entity.setBaseSalary(dto.getBaseSalary());
        entity.setWorkloadSchedule(dto.getWorkloadSchedule());
        entity.setWorkScheduleId(dto.getWorkScheduleId());
        entity.setEmergencyContacts(copyEmergencyContacts(dto.getEmergencyContacts()));
        entity.setContractStartDate(dto.getContractStartDate());
        entity.setContractEndDate(dto.getContractEndDate());
        entity.setProviderCnpj(dto.getProviderCnpj());
        entity.setProviderLegalName(dto.getProviderLegalName());
        entity.setProviderTradeName(dto.getProviderTradeName());
        entity.setRecessEnabled(dto.isRecessEnabled());
        entity.setRecessTotalDaysPerCycle(dto.getRecessTotalDaysPerCycle());
        entity.setRecessCycleMonths(dto.getRecessCycleMonths());
        entity.setRecessEligibilityAfterMonths(dto.getRecessEligibilityAfterMonths());
        entity.setRecessFinancialMode(dto.getRecessFinancialMode());
        entity.setRecessPaidPercentage(dto.getRecessPaidPercentage());
        entity.setRecessAllowSplit(dto.isRecessAllowSplit());
        entity.setRecessMaxSplitPeriods(dto.getRecessMaxSplitPeriods());
        entity.setRecessMinimumSplitDays(dto.getRecessMinimumSplitDays());
        entity.setRecessAdvanceNoticeDays(
                dto.getRecessAdvanceNoticeDays() != null ? dto.getRecessAdvanceNoticeDays() : 0);
        entity.setRecessNotes(dto.getRecessNotes());
        entity.setPhotoObjectId(dto.getPhotoObjectId());
    }

    public AdmissionProcessResponseDto toResponse(AdmissionProcess entity) {
        return toResponse(entity, entity.getAdmissionStatus(), 0, REQUIRED_STEP_COUNT, List.of());
    }

    public AdmissionProcessResponseDto toResponse(AdmissionProcess entity, AdmissionProgress progress) {
        return toResponse(
                entity,
                progress.status(),
                progress.completedStepCount(),
                progress.requiredStepCount(),
                progress.completedStepIds());
    }

    public AdmissionProcessResponseDto toResponse(
            AdmissionProcess entity,
            AdmissionStatusEnum admissionStatus,
            int completedStepCount,
            int requiredStepCount,
            List<String> completedStepIds) {
        return toResponse(entity, admissionStatus, completedStepCount, requiredStepCount, completedStepIds, null);
    }

    public AdmissionProcessResponseDto toResponse(
            AdmissionProcess entity,
            AdmissionStatusEnum admissionStatus,
            int completedStepCount,
            int requiredStepCount,
            List<String> completedStepIds,
            String departmentName) {
        return toResponse(
                entity,
                admissionStatus,
                completedStepCount,
                requiredStepCount,
                completedStepIds,
                departmentName,
                null,
                false,
                false,
                false);
    }

    public AdmissionProcessResponseDto toResponse(
            AdmissionProcess entity,
            AdmissionStatusEnum admissionStatus,
            int completedStepCount,
            int requiredStepCount,
            List<String> completedStepIds,
            String departmentName,
            StatusEnum collaboratorStatus,
            boolean inVacation,
            boolean onLeave,
            boolean onLeaveActive) {
        return AdmissionProcessResponseDto.builder()
                .id(entity.getId())
                .code(entity.getCode())
                .status(entity.getStatus())
                .collaboratorId(entity.getCollaboratorId())
                .collaboratorStatus(collaboratorStatus)
                .photoObjectId(entity.getPhotoObjectId())
                .admissionStatus(admissionStatus)
                .inVacation(inVacation)
                .onLeave(onLeave)
                .onLeaveActive(onLeaveActive)
                .completedStepCount(completedStepCount)
                .requiredStepCount(requiredStepCount)
                .completedStepIds(completedStepIds)
                .startedAt(entity.getStartedAt())
                .completedAt(admissionStatus == AdmissionStatusEnum.COMPLETED ? entity.getCompletedAt() : null)
                .notes(entity.getNotes())
                .fullName(entity.getFullName())
                .socialName(entity.getSocialName())
                .cpf(entity.getCpf())
                .rg(entity.getRg())
                .rgIssuer(entity.getRgIssuer())
                .rgStateCode(entity.getRgStateCode())
                .birthDate(entity.getBirthDate())
                .gender(entity.getGender())
                .maritalStatus(entity.getMaritalStatus())
                .motherName(entity.getMotherName())
                .fatherName(entity.getFatherName())
                .nationality(entity.getNationality())
                .pisPasep(entity.getPisPasep())
                .email(entity.getEmail())
                .phone(entity.getPhone())
                .zipCode(entity.getZipCode())
                .street(entity.getStreet())
                .number(entity.getNumber())
                .complement(entity.getComplement())
                .district(entity.getDistrict())
                .cityId(entity.getCity() != null ? entity.getCity().getId() : null)
                .cityName(entity.getCity() != null ? entity.getCity().getName() : null)
                .stateId(entity.getState() != null ? entity.getState().getId() : null)
                .stateCode(entity.getState() != null ? entity.getState().getAbbreviation() : null)
                .stateName(entity.getState() != null ? entity.getState().getName() : null)
                .expectedStartDate(entity.getExpectedStartDate())
                .companyId(entity.getCompanyId())
                .departmentId(entity.getDepartmentId())
                .departmentName(departmentName)
                .jobPositionId(entity.getJobPositionId())
                .contractType(entity.getContractType())
                .baseSalary(entity.getBaseSalary())
                .workloadSchedule(entity.getWorkloadSchedule())
                .workScheduleId(entity.getWorkScheduleId())
                .emergencyContacts(copyEmergencyContacts(entity.getEmergencyContacts()))
                .contractStartDate(entity.getContractStartDate())
                .contractEndDate(entity.getContractEndDate())
                .providerCnpj(entity.getProviderCnpj())
                .providerLegalName(entity.getProviderLegalName())
                .providerTradeName(entity.getProviderTradeName())
                .recessEnabled(entity.isRecessEnabled())
                .recessTotalDaysPerCycle(entity.getRecessTotalDaysPerCycle())
                .recessCycleMonths(entity.getRecessCycleMonths())
                .recessEligibilityAfterMonths(entity.getRecessEligibilityAfterMonths())
                .recessFinancialMode(entity.getRecessFinancialMode())
                .recessPaidPercentage(entity.getRecessPaidPercentage())
                .recessAllowSplit(entity.isRecessAllowSplit())
                .recessMaxSplitPeriods(entity.getRecessMaxSplitPeriods())
                .recessMinimumSplitDays(entity.getRecessMinimumSplitDays())
                .recessAdvanceNoticeDays(entity.getRecessAdvanceNoticeDays())
                .recessNotes(entity.getRecessNotes())
                .createdAt(entity.getCreatedAt())
                .updatedAt(entity.getUpdatedAt())
                .build();
    }

    private static List<AdmissionEmergencyContactDto> copyEmergencyContacts(
            List<AdmissionEmergencyContactDto> contacts) {
        if (CollectionUtils.isEmpty(contacts)) {
            return new ArrayList<>();
        }
        return contacts.stream()
                .map(contact -> AdmissionEmergencyContactDto.builder()
                        .name(contact.getName())
                        .phone(contact.getPhone())
                        .relationship(contact.getRelationship())
                        .build())
                .toList();
    }
}

File: gommo-frontend/src/modules/rh/person/collaborators/admission/lib/admission-process.mapper.ts
import type {
    AdmissionEmergencyContact,
    AdmissionProcess,
    AdmissionProcessCreateDto,
} from "@/modules/rh/person/collaborators/admission/dto/admission-process.dto";
import { normalizeEmergencyContacts } from "@/modules/rh/person/collaborators/admission/lib/admission-emergency-contacts.util";
import { digitsOnly } from "@/shared/lib/input/digits";

const ADMISSION_STEP_IDS = [
    "dados-basicos",
    "contatos-emergencia",
    "endereco",
    "documentos",
    "vinculo",
    "contrato",
    "observacoes",
];

function toDateInput(value?: string): string {
    return value?.slice(0, 10) ?? "";
}

export function admissionprocessToFormDto(entity: AdmissionProcess): AdmissionProcessCreateDto {
    return {
        admissionStatus: entity.admissionStatus,
        startedAt: toDateInput(entity.startedAt),
        notes: entity.notes ?? "",
        photoObjectId: entity.photoObjectId ?? "",
        fullName: entity.fullName ?? "",
        socialName: entity.socialName ?? "",
        cpf: entity.cpf ?? "",
        rg: entity.rg ?? "",
        rgIssuer: entity.rgIssuer ?? "",
        rgStateCode: entity.rgStateCode ?? "",
        birthDate: toDateInput(entity.birthDate),
        gender: entity.gender,
        maritalStatus: entity.maritalStatus,
        motherName: entity.motherName ?? "",
        fatherName: entity.fatherName ?? "",
        nationality: entity.nationality ?? "Brasileiro",
        pisPasep: entity.pisPasep ?? "",
        email: entity.email ?? "",
        phone: entity.phone ?? "",
        zipCode: entity.zipCode ?? "",
        street: entity.street ?? "",
        number: entity.number ?? "",
        complement: entity.complement ?? "",
        district: entity.district ?? "",
        cityId: entity.cityId ?? "",
        stateId: entity.stateId ?? "",
        expectedStartDate: toDateInput(entity.expectedStartDate),
        companyId: entity.companyId ?? "",
        departmentId: entity.departmentId ?? "",
        jobPositionId: entity.jobPositionId ?? "",
        contractType: entity.contractType ?? "CLT",
        baseSalary: entity.baseSalary != null ? String(entity.baseSalary) : "",
        workloadSchedule: entity.workloadSchedule ?? "",
        workScheduleId: entity.workScheduleId ?? "",
        emergencyContacts: normalizeEmergencyContacts(entity.emergencyContacts),
        contractStartDate: toDateInput(entity.contractStartDate),
        contractEndDate: toDateInput(entity.contractEndDate),
        providerCnpj: entity.providerCnpj ?? "",
        providerLegalName: entity.providerLegalName ?? "",
        providerTradeName: entity.providerTradeName ?? "",
        recessEnabled: entity.recessEnabled ?? false,
        recessTotalDaysPerCycle: entity.recessTotalDaysPerCycle ?? "",
        recessCycleMonths: entity.recessCycleMonths ?? "",
        recessEligibilityAfterMonths: entity.recessEligibilityAfterMonths ?? "",
        recessFinancialMode: entity.recessFinancialMode,
        recessPaidPercentage: entity.recessPaidPercentage ?? "",
        recessAllowSplit: entity.recessAllowSplit ?? false,
        recessMaxSplitPeriods: entity.recessMaxSplitPeriods ?? "",
        recessMinimumSplitDays: entity.recessMinimumSplitDays ?? "",
        recessAdvanceNoticeDays: entity.recessAdvanceNoticeDays ?? 0,
        recessNotes: entity.recessNotes ?? "",
    };
}

export const emptyAdmissionProcessForm = (): AdmissionProcessCreateDto => ({
    admissionStatus: "IN_PROGRESS",
    startedAt: new Date().toISOString().slice(0, 10),
    notes: "",
    fullName: "",
    socialName: "",
    cpf: "",
    rg: "",
    rgIssuer: "",
    rgStateCode: "",
    birthDate: "",
    nationality: "Brasileiro",
    expectedStartDate: "",
    contractType: "CLT",
    companyId: "",
    departmentId: "",
    jobPositionId: "",
    cityId: "",
    stateId: "",
    workloadSchedule: "",
    workScheduleId: "",
    emergencyContacts: normalizeEmergencyContacts(),
    contractStartDate: "",
    contractEndDate: "",
    providerCnpj: "",
    providerLegalName: "",
    providerTradeName: "",
    recessEnabled: false,
    recessTotalDaysPerCycle: "",
    recessCycleMonths: "",
    recessEligibilityAfterMonths: "",
    recessPaidPercentage: "",
    recessAllowSplit: false,
    recessMaxSplitPeriods: "",
    recessMinimumSplitDays: "",
    recessAdvanceNoticeDays: 0,
    recessNotes: "",
});

export function admissionFormToPayload(form: AdmissionProcessCreateDto): AdmissionProcessCreateDto {
    const emergencyContacts = sanitizeEmergencyContacts(form.emergencyContacts);
    return {
        ...form,
        cpf: digitsOnly(form.cpf),
        phone: form.phone ? digitsOnly(form.phone) : undefined,
        zipCode: form.zipCode ? digitsOnly(form.zipCode) : undefined,
        emergencyContacts,
        cityId: form.cityId?.trim() || undefined,
        stateId: form.stateId?.trim() || undefined,
        companyId: form.companyId?.trim() || undefined,
        departmentId: form.departmentId?.trim() || undefined,
        jobPositionId: form.jobPositionId?.trim() || undefined,
        socialName: form.socialName?.trim() || undefined,
        rg: form.rg?.trim() ? digitsOnly(form.rg).slice(0, 7) : undefined,
        rgIssuer: form.rgIssuer?.trim() || undefined,
        rgStateCode: form.rgStateCode?.trim().toUpperCase() || undefined,
        motherName: form.motherName?.trim() || undefined,
        fatherName: form.fatherName?.trim() || undefined,
        pisPasep: form.pisPasep?.trim() || undefined,
        email: form.email?.trim() || undefined,
        notes: form.notes?.trim() || undefined,
        photoObjectId: form.photoObjectId?.trim() || undefined,
        workloadSchedule: form.workloadSchedule?.trim() || undefined,
        workScheduleId: form.workScheduleId?.trim() || undefined,
        contractStartDate: form.contractStartDate || undefined,
        contractEndDate: form.contractEndDate || undefined,
        providerCnpj: form.providerCnpj ? digitsOnly(form.providerCnpj) : undefined,
        providerLegalName: form.providerLegalName?.trim() || undefined,
        providerTradeName: form.providerTradeName?.trim() || undefined,
        baseSalary: form.baseSalary != null && form.baseSalary !== "" ? Number(form.baseSalary) : undefined,
        recessTotalDaysPerCycle: toOptionalNumber(form.recessTotalDaysPerCycle),
        recessCycleMonths: toOptionalNumber(form.recessCycleMonths),
        recessEligibilityAfterMonths: toOptionalNumber(form.recessEligibilityAfterMonths),
        recessPaidPercentage: toOptionalNumber(form.recessPaidPercentage),
        recessMaxSplitPeriods: toOptionalNumber(form.recessMaxSplitPeriods),
        recessMinimumSplitDays: toOptionalNumber(form.recessMinimumSplitDays),
        recessAdvanceNoticeDays: toOptionalNumber(form.recessAdvanceNoticeDays) ?? 0,
        recessNotes: form.recessNotes?.trim() || undefined,
    };
}

function toOptionalNumber(value?: string | number): number | undefined {
    return value == null || value === "" ? undefined : Number(value);
}

function sanitizeEmergencyContacts(contacts?: AdmissionEmergencyContact[]): AdmissionEmergencyContact[] {
    return (contacts ?? [])
        .map((contact) => ({
            name: contact.name?.trim() ?? "",
            phone: contact.phone ? digitsOnly(contact.phone) : "",
            relationship: contact.relationship?.trim() || undefined,
        }))
        .filter((contact) => contact.name || contact.phone || contact.relationship);
}

export { ADMISSION_STEP_IDS };
